File: downloads/code/LeetCode/Java/Shuffle-an-Array.java
public class Solution {

    int[] nums;
    Random random;
    public Solution(int[] nums) {
        this.nums = nums;
        this.random = new Random();
    }
    
    /** Resets the array to its original configuration and return it. */
    public int[] reset() {
        return this.nums;
    }
    
    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
        int[] ret = this.nums.clone();
        for(int j=1;j<this.nums.length;j++){
            int i = random.nextInt(j+1);
            swap(ret, i,j);
        }
        return ret;
    }

    public void swap (int[] nums, int a, int b){
        int tmp = nums[a];
        nums[a] = nums[b];
        nums[b] = tmp;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */

File: downloads/code/LeetCode/Python/Shuffle-an-Array.py
class Solution(object):

    def __init__(self, nums):
        """
        
        :type nums: List[int]
        :type size: int
        """
        self.orig = nums[:]
        self.shuffled_nums = nums[:]

    def reset(self):
        """
        Resets the array to its original configuration and return it.
        :rtype: List[int]
        """
        self.shuffled_nums = self.orig[:]
        return self.orig


    def shuffle(self):
        """
        Returns a random shuffling of the array.
        :rtype: List[int]
        """
        n = len(self.shuffled_nums)
        for i in range(n):
            newindex = random.randint(0,i)
            self.shuffled_nums[i],self.shuffled_nums[newindex] = self.shuffled_nums[newindex],self.shuffled_nums[i]
	return self.shuffled_nums

# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()
